tests/port-forwarding-server.spec.ts

```diff
@@ -0,0 +1,133 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import childProcess from 'child_process';
+import http from 'http';
+import path from 'path';
+import net from 'net';
+
+import { contextTest, expect } from './config/browserTest';
+import { PlaywrightClient } from '../lib/remote/playwrightClient';
+import type { Page, BrowserServer } from '..';
+
+class OutOfProcessPlaywrightServer {
+  private _driverProcess: childProcess.ChildProcess;
+  private _receivedPortPromise: Promise<string>;
+
+  constructor(port: number, proxyPort: number) {
+    this._driverProcess = childProcess.fork(path.join(__dirname, '..', 'lib', 'cli', 'cli.js'), ['run-server', port.toString()], {
+      stdio: 'pipe',
+      detached: true,
+      env: {
+        ...process.env,
+        PW_SOCKS_PROXY_PORT: String(proxyPort)
+      }
+    });
+    this._driverProcess.unref();
+    this._receivedPortPromise = new Promise<string>((resolve, reject) => {
+      this._driverProcess.stdout.on('data', (data: Buffer) => {
+        const prefix = 'Listening on ';
+        const line = data.toString();
+        if (line.startsWith(prefix))
+          resolve(line.substr(prefix.length));
+      });
+      this._driverProcess.on('exit', () => reject());
+    });
+  }
+  async kill() {
+    const waitForExit = new Promise<void>(resolve =>  this._driverProcess.on('exit', () => resolve()));
+    this._driverProcess.kill('SIGKILL');
+    await waitForExit;
+  }
+  public async wsEndpoint(): Promise<string> {
+    return await this._receivedPortPromise;
+  }
+}
+
+const it = contextTest.extend<{ pageFactory: (redirectPortForTest?: number) => Promise<Page> }>({
+  pageFactory: async ({ browserName, browserOptions }, run, testInfo) => {
+    const browserServers: BrowserServer[] = [];
+    const playwrightServers: OutOfProcessPlaywrightServer[] = [];
+    await run(async (redirectPortForTest?: number): Promise<Page> => {
+      const server = new OutOfProcessPlaywrightServer(0, 3200 + testInfo.workerIndex);
+      playwrightServers.push(server);
+      const service = await PlaywrightClient.connect({
+        wsEndpoint: await server.wsEndpoint(),
+      });
+      const playwright = service.playwright();
+      playwright._enablePortForwarding(redirectPortForTest);
+      const browser = await playwright[browserName].launch(browserOptions);
+      return await browser.newPage();
+    });
+    for (const browserServer of browserServers)
+      await browserServer.close();
+    for (const playwrightServer of playwrightServers)
+      await playwrightServer.kill();
+  },
+});
+
+it.fixme(({ platform, browserName }) => platform === 'darwin' && browserName === 'webkit');
```

Comment:
for reference, lets add a comment to it: https://github.com/mxschmitt/cfnetwork-localhost-proxy

Reply:
I'll look at it.